File: src/dataset/go/say_Exercism_go_ThomasZumsteg/openAI/ChatGPT4_say_v2.go
package say

import (
	"strings"
)

// Predefined constants for number-to-words conversion
var (
	ones = []string{"", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
	teens = []string{"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"}
	tens = []string{"", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}
	powers = []string{"", "thousand", "million", "billion", "trillion", "quadrillion", "quintillion"}
)

// Say converts a number to its English phrase representation
func Say(num uint64) string {
	if num == 0 {
		return "zero"
	}

	var result strings.Builder
	powerIndex := 0

	for num > 0 {
		group := num % 1000
		if group > 0 {
			groupWords := sayPower(group)
			if powerIndex > 0 {
				result.WriteString(powers[powerIndex] + " ")
			}
			result.WriteString(groupWords)
		}
		num /= 1000
		powerIndex++
	}

	words := result.String()
	return strings.TrimSpace(reverseWords(words))
}

// sayPower converts a number (1-999) to its English phrase representation
func sayPower(num uint64) string {
	var result strings.Builder
	hundred, ten, one := num/100, (num%100)/10, num%10

	if hundred > 0 {
		result.WriteString(ones[hundred] + " hundred ")
	}

	if ten == 1 {
		result.WriteString(teens[one] + " ")
	} else {
		if ten > 0 {
			result.WriteString(tens[ten] + " ")
		}
		if one > 0 {
			result.WriteString(ones[one] + " ")
		}
	}

	return result.String()
}

// reverseWords reverses the order of words in a string
func reverseWords(s string) string {
	words := strings.Fields(s)
	for i, j := 0, len(words)-1; i < j; i, j = i+1, j-1 {
		words[i], words[j] = words[j], words[i]
	}
	return strings.Join(words, " ")
}

File: src/dataset/Java/say_Exercism_java_rabestro/openAI/ChatGPT4_Say_v2.java
public final class Say {
    public static final long MAXIMUM_PRONOUNCEABLE_NUMBER = 999_999_999_999L;
    private static final String[] BELOW_TWENTY = {
        "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", 
        "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", 
        "seventeen", "eighteen", "nineteen"
    };
    private static final String[] TENS = {
        "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
    };
    private static final String[] MAGNITUDES = {"", "thousand", "million", "billion"};

    public String say(long number) {
        if (number < 0 || number > MAXIMUM_PRONOUNCEABLE_NUMBER) {
            throw new IllegalArgumentException("Number must be between 0 and " + MAXIMUM_PRONOUNCEABLE_NUMBER + ".");
        }
        if (number == 0) return BELOW_TWENTY[0];
        return sayNumber(number).trim();
    }

    private String sayNumber(long number) {
        StringBuilder result = new StringBuilder();
        int magnitudeIndex = 0;

        while (number > 0) {
            int chunk = (int) (number % 1_000);
            if (chunk > 0) {
                result.insert(0, sayChunk(chunk) + (magnitudeIndex > 0 ? " " + MAGNITUDES[magnitudeIndex] : "") + " ");
            }
            number /= 1_000;
            magnitudeIndex++;
        }

        return result.toString().trim();
    }

    private String sayChunk(int number) {
        StringBuilder chunkResult = new StringBuilder();

        if (number >= 100) {
            chunkResult.append(BELOW_TWENTY[number / 100]).append(" hundred");
            number %= 100;
            if (number > 0) chunkResult.append(" ");
        }

        if (number >= 20) {
            chunkResult.append(TENS[number / 10]);
            number %= 10;
            if (number > 0) chunkResult.append("-");
        }

        if (number > 0) {
            chunkResult.append(BELOW_TWENTY[number]);
        }

        return chunkResult.toString();
    }
}